src/main/java/org/opensearch/security/privileges/UserAttributes.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+package org.opensearch.security.privileges;
+
+import java.util.Map;
+import java.util.Set;
+
+import com.google.common.base.Joiner;
+import com.google.common.collect.Iterables;
+
+import org.opensearch.security.user.User;
+
+/**
+ * Support for interpolating user attributes used in index patterns and DLS queries.
+ *
+ * This code was moved over from ConfigModelV7.
+ */
+public class UserAttributes {
+    public static String replaceProperties(String orig, User user) {
+
+        if (user == null || orig == null) {
+            return orig;
+        }
+
+        orig = orig.replace("${user.name}", user.getName()).replace("${user_name}", user.getName());
+        orig = replaceRoles(orig, user);
+        orig = replaceSecurityRoles(orig, user);
+        for (Map.Entry<String, String> entry : user.getCustomAttributesMap().entrySet()) {
+            if (entry == null || entry.getKey() == null || entry.getValue() == null) {
+                continue;
+            }
+            orig = orig.replace("${" + entry.getKey() + "}", entry.getValue());
+            orig = orig.replace("${" + entry.getKey().replace('.', '_') + "}", entry.getValue());
+        }
+        return orig;
+    }
+
+    private static String replaceRoles(final String orig, final User user) {
+        String retVal = orig;
+        if (orig.contains("${user.roles}") || orig.contains("${user_roles}")) {
+            final String commaSeparatedRoles = toQuotedCommaSeparatedString(user.getRoles());
+            retVal = orig.replace("${user.roles}", commaSeparatedRoles).replace("${user_roles}", commaSeparatedRoles);
```

Comment:
TODO: It would be good to check if multiple String.replace() calls can be merged into single StringBuilder.replace() and measure performance for our use-case. This should be possible once Security plugin has micro-benchmark functionality using JMH as proposed in https://github.com/opensearch-project/security/pull/4380

Reply:
IMHO, it would be best to redesign the whole attributes handling to use pre-parsed templates instead of these simple string replacement operations.

This would have a couple of advantages:

- Performance
- One could detect the use of unknown attributes and log errors. At the moment, unknown attributes are just left as such in the resulting string.
- Possibility for support for multi-valued attributes. At the moment, multi-valued attributes are not really supported. If one wants to achieve that, some pre-fabricated strings in the correct syntax (JSON or index pattern) need to be defined in the auth backend.

This, however, should probably done in a separate effort.

Reply:
Lets not block merging for these follow ups